File: TcpListStreamDemo/tcpliststream.cpp
#include "tcpliststream.h"

TcpListStream::TcpListStream()
{
}

void TcpListStream::setSocket(QIODevice *socket)
{
    if (mSocket)
        disconnect(mSocket, 0, this, 0);

    if (socket)
        connect(socket, SIGNAL(readyRead()), SLOT(socketReadyRead()));

    mSocket = socket;
}

QIODevice *TcpListStream::getSocket()
{
    return mSocket;
}

void TcpListStream::writeList(const QStringList &list)
{
    const int listSize = list.size();

    QVector<int> headerData(listSize);

    //Concatinate the strings and grab the header positions
    QString message;
    for (int i = 0; i < listSize; i++) {
        int termStartPos = message.size();
        if (termStartPos != 0)
            headerData[i - 1] = termStartPos;

        message += list.at(i);
    }

    //Set the last header number to the total message size
    QByteArray messageBytes = message.toUtf8();
    const int totalMessageSize = messageBytes.size();
    headerData[listSize - 1] = totalMessageSize;

    //Convert the list of ints into a QString header
    QString header;
    header.reserve(listSize * 2);
    for (int i = 0; i < listSize; i++) {
        header += QString::number(headerData.at(i));

        if (i != listSize - 1)
            header += QChar('.');
        else
            header += QChar(' ');
    }

    //Build the final message
    QByteArray finalMessage;
    finalMessage.append(header);
    finalMessage.append(messageBytes);

    mSocket->write(finalMessage);
}

QStringList TcpListStream::readList()
{
    return mReceivedLists.dequeue();
}

QList<QStringList> TcpListStream::readAll()
{
    QList<QStringList> lists;

    for (int i = 0; i < mReceivedLists.size(); i++) {
        lists.append(readList());
    }

    return lists;
}

void TcpListStream::socketReadyRead()
{
    mSpaceSearchPos = mBuffer.size();
    mBuffer.append(mSocket->readAll());

    while (true) {
        if (mHeader.isEmpty()) {
            mSpaceLocation = mBuffer.indexOf(' ', mSpaceSearchPos);

            //Levae if we could not find a space
            if (mSpaceLocation == -1)
                return;

            if (!fetchHeader()) {
                resetReadData(-1);
                return;
            }

            //Figure out the length of the entire message
            mLength = mSpaceLocation + mHeader.back() + 1;
        }

        //Leave if we don't have all of the message yet
        if (mBuffer.size() < mLength)
            return;

        const QByteArray messageBytes = mBuffer.mid(mSpaceLocation + 1, mHeader.back());

        const QString message = QString::fromUtf8(messageBytes);

        QStringList list;
        list.reserve(mHeader.size());

        for (int i = 0; i < mHeader.size(); i++) {
            int startPos, endPos, length;

            if (i == 0)
                startPos = 0;
            else
                startPos = mHeader.at(i - 1);

            if (i == mHeader.size() - 1)
                endPos = message.size();
            else
                endPos = mHeader.at(i);

            length = endPos - startPos;

            const QString data = message.mid(startPos, length);
            list.append(data);
        }

        //Empty the buffer and header
        resetReadData(mLength);

        mReceivedLists.enqueue(list);
        emit readyRead();

        if (mBuffer.isEmpty())
            return;
    }
}

bool TcpListStream::fetchHeader()
{
    const QString header = QString::fromUtf8(mBuffer.left(mSpaceLocation));
    const QStringList headerData = header.split('.');
    const int headerDataSize = headerData.size();

    if (headerDataSize == 0)
        return false;

    mHeader.resize(headerDataSize);

    for (int i = 0; i < headerDataSize; i++) {
        int position = headerData.at(i).toInt();
        if (position != 0)
            mHeader[i] = position;
        else
            return false;
    }

    return true;
}

void TcpListStream::resetReadData(int bufferLength)
{
    mBuffer.remove(0, bufferLength);
    mHeader.clear();
    mHeader.squeeze();
    mSpaceSearchPos = 0;
}

File: java TcpListStreamDemo/src/TcpListStream.java
import java.nio.charset.StandardCharsets;
import java.util.ArrayDeque;
import java.util.List;
import java.util.Queue;

public class TcpListStream implements ITcpListStream, IBasicIODeviceListener {

    private BasicTcpSocket mSocket = null;

    private Queue<List<String>> mReceivedLists = new ArrayDeque<>();

    public void setSocket(BasicTcpSocket socket) {
        if (mSocket != null)
            mSocket.removeListener(this);

        if (socket!= null)
            socket.addListener(this);

        mSocket = socket;
    }

    public BasicTcpSocket getSocket() {
        return mSocket;
    }

    public void readyRead() {

    }

    public void writeList(List<String> list) {

        final int listSize = list.size();

        int[] headerData = new int[listSize];

        //Concatinate the strings and grab the header positions
        StringBuilder message = new StringBuilder();
        for (int i = 0; i < listSize; i++) {
            final int termStartPos = message.length();
            if (termStartPos != 0)
                headerData[i - 1] = termStartPos;

            message.append(list.get(i));
        }

        //Set the last header number to the total message size
        byte[] messageBytes = message.toString().getBytes(StandardCharsets.UTF_8);
        final int totalMessageSize = messageBytes.length;
        headerData[listSize - 1] = totalMessageSize;

        //Convert the list of ints into a String header
        StringBuilder header = new StringBuilder();
        for (int i = 0; i < listSize; i++) {
            header.append(headerData[i]);

            if (i != listSize - 1)
                header.append('.');
            else
                header.append(' ');
        }

        byte[] headerBytes = header.toString().getBytes(StandardCharsets.UTF_8);

        byte[] finalMessage = new byte[headerBytes.length + messageBytes.length];
        System.arraycopy(headerBytes, 0, finalMessage, 0, headerBytes.length);
        System.arraycopy(messageBytes, 0, finalMessage, headerBytes.length, messageBytes.length);

        mSocket.write(finalMessage);
    }

    public List<String> readList() {
        return null;
    }

    public List<List<String>> readAll() {
        return null;
    }

    //readyRead signal
    public void addListener(IBasicIODeviceListener listener) {

    }

    public void removeListener(IBasicIODeviceListener listener) {

    }
}
